docs/src/main/asciidoc/rest-client-reactive.adoc

```diff
@@ -591,6 +591,37 @@ To specify a value for `${header.value}`, simply put the following in your `appl
 header.value=value of the header
 ----
 
+Also, there is a reactive flavor of `ClientHeadersFactory` that allows doing blocking operations. For example:
+
+[source, java]
+----
+package org.acme.rest.client;
+
+import org.eclipse.microprofile.rest.client.ext.ClientHeadersFactory;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.ws.rs.core.MultivaluedHashMap;
+import javax.ws.rs.core.MultivaluedMap;
+import java.util.UUID;
+
+@ApplicationScoped
+public class GetTokenReactiveClientHeadersFactory extends ReactiveClientHeadersFactory {
+
+    @Inject
+    Service service;
+
+    @Override
+    public Uni<MultivaluedMap<String, String>> getHeaders(MultivaluedMap<String, String> incomingHeaders) {
+        return Uni.createFrom().item(() -> {
+            MultivaluedHashMap<String, String> newHeaders = new MultivaluedHashMap<>();
+            // perform blocking call
+            newHeaders.add(HEADER_NAME, service.getToken());
```

Comment:
Hi,
I am sorry that I kind of hijack this already merged PR, but IMHO the `ReactiveClientHeadersFactory` does not work as expected.
Why has the `service.getToken()` to be a blocking call?
In my case I wanted to call the auth server where I obtain a token also in a reactive manner by using an reactive rest client.
Therefore my code would look like this:
```
    override fun getHeaders(incomingHeaders: MultivaluedMap<String, String>?): Uni<MultivaluedMap<String, String>> {
        // Utilizing the reactive mutiny api :-), which unfortunately does not work....
        return clientTokenFetcher.getClientToken().onItem().transform {
            MultivaluedHashMap<String, String>().apply {
                add(HttpHeaders.CACHE_CONTROL, "no-cache")
                add(HttpHeaders.CONTENT_TYPE, "text/plain")
                add(HttpHeaders.AUTHORIZATION, it.authHeader)
            }
        }
    }
``` 

Unfortunately this does not work, while this is working:

```
    override fun getHeaders(incomingHeaders: MultivaluedMap<String, String>?): Uni<MultivaluedMap<String, String>> {
        return Uni.createFrom().item(
            MultivaluedHashMap<String, String>().apply {
                add(HttpHeaders.CACHE_CONTROL, "no-cache")
                add(HttpHeaders.CONTENT_TYPE, "text/plain")
                // Make this a blocking call ;-(
                add(HttpHeaders.AUTHORIZATION, clientTokenFetcher.getClientToken().await().atMost(Duration.ofMillis(500)).authHeader)
            }
        )
    }
```

Can someone please tell me how to stay fully reactive and make it work without the blocking within the Uni?

The code above is written in kotlin, but I could also convert it into Java if that helps you...

Reply:
The `service.getToken()` does not have to be a blocking call. This example was only intended to explain how to inject headers from a blocking service to Resteasy Rest Client. 

Also, I tried to reproduce your scenario and it worked for me:

```java
    @ApplicationScoped
    public static class Service {
        public Uni<String> getValue() {
            return Uni.createFrom().item(HEADER_VALUE);
        }
    }

    @RegisterClientHeaders(CustomReactiveClientHeadersFactory.class)
    public interface Client {
        @GET
        String getWithHeader();
    }

    public static class CustomReactiveClientHeadersFactory extends ReactiveClientHeadersFactory {

        @Inject
        Service service;

        @Override
        public Uni<MultivaluedMap<String, String>> getHeaders(MultivaluedMap<String, String> incomingHeaders) {
            return service.getValue().onItem().transform(value -> {
                MultivaluedHashMap<String, String> newHeaders = new MultivaluedHashMap<>();
                newHeaders.add(HEADER_NAME, value);
                return newHeaders;
            });
        }
    }
```

This is verified using this test https://github.com/quarkusio/quarkus/blob/main/extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/java/io/quarkus/rest/client/reactive/headers/ReactiveClientHeadersFromProviderTest.java

Therefore, it would be good to open an issue with a reproducer, so we can exactly see what is going on.

Reply:
Hi @Sgitario ,

thanks for your quick response:
Here you can find an example which reproduces my issue:
https://github.com/SimonScholz/ReactiveClientHeadersFactory/blob/b1a7a07c13afa9168b8897653fa27ce0f4cf3be3/src/test/kotlin/io/github/simonscholz/ReactiveResourceTest.kt#L82

In this example I'd use a reactive rest client in order to obtain a token, which is then used by the `BeverageRequestHeaderFactory` to generate a proper Authorization header.
The `ReactiveResourceTest` shows that this does not work.
I'll also create an issue later on, but wanted to inform you about my example in advance.

Thanks and regards,

Simon

Reply:
Hi @Sgitario,

FYI, I've now finally created an issue here: https://github.com/quarkusio/quarkus/issues/24153

Please let me know, if I can do more or if you have any questions.